File: Pf/nexusToken.c
#include <stdio.h>
#include "pftypes.h"
#include "nexusToken.h"

void handleComment(nexusToken *nt);
void printComment(nexusToken *nt);
void getComment(nexusToken *nt);
void skipComment(nexusToken *nt);
void getQuotedStuff(nexusToken *nt);
void getWord(nexusToken *nt);

void nextTok(nexusToken *nt)
{
	int c;

	//printf("nextTok() here.  nt->filePtr=%li, fseek=%li\n", (long int)nt->filePtr, ftell(nt->filePtr));
	//printf("start of nextTok() here.  nt->tokLen=%i, embeddedCommentLen=%i, savedCommentLen=%i\n", nt->tokLen[0], nt->embeddedCommentLen[0], nt->savedCommentLen[0]);
	//printf("c nextTok() here.  nt->max=%i\n", nt->max[0]);
	while(1) {
		c = fgetc(nt->filePtr);
		//printf("c nextTok()  got c=%c\n", c);
		if(c == EOF) {
			//printf("nextTok() got EOF\n");
			break;
		}
		if(nt->getLineEndings[0]) {
			if(c == '\n' || c == '\r') {
				nt->tokLen[0] = 1;
				nt->tok[0] = '\n';  // note only returns \n
				break;
			}
		}
		if(c == ' ' || c == '\n' || c == '\t' || c == '\r') {
			//printf("c nextTok() got space\n");
			continue;
		}
		else if(c == '[') {
			handleComment(nt);
			if(nt->tokLen[0]) {
				break;
			}
			else {
				continue;
			}
		}
		else if(c == '\'') {
			fseek(nt->filePtr, -1, 1);
			getQuotedStuff(nt);
			break;
		}
		else if(c == '(' || c == ')' || c == '[' || c == ']' || c == '{' || 
				c == '}' || c == '\\' || c == '/' || c == ',' || c == ';' || 
				c == ':' || c == '=' || c == '*' || c == '\'' || c == '"' || 
				c == '`' || c == '+' || c == '-' || c == '<' || c == '>') {
			nt->tok[0] = c;
			nt->tokLen[0] = 1;
			break;
		}
		else {
			//printf("c nextTok() about to getWord(). nt->tokLen = %i\n", nt->tokLen[0]);
			fseek(nt->filePtr, -1, 1);
			getWord(nt);
			//printf("c nextTok() after getWord(). nt->tokLen = %i\n", nt->tokLen[0]);
			break;
		}
	}
	//printf("\n");

}

void handleComment(nexusToken *nt)
{
	long int   commentStartPos;
	int       c, c2;
	int        i;
	char       p4[3];

	// We are here having read in the opening '[' character.
	commentStartPos = ftell(nt->filePtr) - 1;
	// This next would be the char following the '['
	c = fgetc(nt->filePtr);
	//printf("c handleComment() c = %c\n", c);
	if(c == EOF) {
		printf("nextToken(), handleComment().  Reached the end of the file while still in a comment.\n");
		exit(1);
	}
	if(c == ']') {
		return;
	}
	if(c == '!' && nt->writeVisibleComments[0]) {
		fseek(nt->filePtr, commentStartPos, 0);
		printComment(nt);
		return;
	}
	else if(c == '&') {
		//printf("c handleComment()  nt->getAllCommandComments[0] = %i\n", nt->getAllCommandComments[0]);
		if(nt->getAllCommandComments[0]) {
			fseek(nt->filePtr, commentStartPos, 0);
			getComment(nt);
			return;
		}
		c2 = fgetc(nt->filePtr);
		if(c2 == EOF) {
			printf("nextToken(), handleComment().  Reached the end of the file while still in a comment.\n");
			exit(1);
		}
		if(c == '&' && (c2 == 'W' || c2 == 'w') && nt->getWeightCommandComments[0]) {
			fseek(nt->filePtr, commentStartPos, 0);
			getComment(nt);
			return;
		}
		else if(c2 == '&' && nt->getP4CommandComments[0]) {
			p4[0] = 'p';
			p4[1] = '4';
			p4[2] = ' ';
			for(i = 0; i < 3; i++) {
				c2 = fgetc(nt->filePtr);
				if(c2 == EOF) {
					printf("nextToken(), handleComment().  Reached the end of the file while still in a comment.\n");
					exit(1);
				}
				if(c2 != p4[i]) {
					fseek(nt->filePtr, commentStartPos, 0);
					skipComment(nt);
					return;
				}
			}
			fseek(nt->filePtr, commentStartPos, 0);
			getComment(nt);
			return;
		}
		else {
			fseek(nt->filePtr, commentStartPos, 0);
			skipComment(nt);
			return;
		}			
	}
	else {
		fseek(nt->filePtr, commentStartPos, 0);
		skipComment(nt);
		return;
	}
	
}

void printComment(nexusToken *nt)
{
	int   level = 0;
	int   c;

	while(1) {
		c = fgetc(nt->filePtr);
		if(c == EOF) {
			printf("nextToken(), printComment().  Reached the end of the file while still in a comment.\n");
			exit(1);
		}
		printf("%c", c);
		if(c == '[') {
			level++;
		}
		else if(c == ']') {
			level--;
		}
		if(level == 0) {
			printf("\n");
			return;
		}
	}
	
}


void skipComment(nexusToken *nt)
{
	int   level = 0;
	int   c;

	while(1) {
		c = fgetc(nt->filePtr);
		if(c == EOF) {
			printf("nextToken(), skipComment().  Reached the end of the file while still in a comment.\n");
			exit(1);
		}
		if(c == '[') {
			level++;
		}
		else if(c == ']') {
			level--;
		}
		if(level == 0) {
			return;
		}
	}
}

void getComment(nexusToken *nt)
{
	int   level = 0;
	//long int startPos = ftell(nt->filePtr);
	int   c;

	if(nt->tokLen[0]) {
		while(1) {
			c = fgetc(nt->filePtr);
			if(c == EOF) {
				printf("nextToken(), getComment().  Reached the end of the file while still in a comment.\n");
				exit(1);
			}
			nt->embeddedComment[nt->embeddedCommentLen[0]] = c;
			nt->embeddedCommentLen[0]++;
			//printf("   getComment().  embeddedCommentLen[0] is %i\n", nt->embeddedCommentLen[0]);

			if(c == '[') {
				level++;
			}
			else if(c == ']') {
				level--;
			}
			if(level == 0) {
				return;
			}
		}
	}
	else {
		while(1) {
			c = fgetc(nt->filePtr);
			if(c == EOF) {
				printf("nextToken(), getComment().  Reached the end of the file while still in a comment.\n");
				exit(1);
			}
			nt->tok[nt->tokLen[0]] = c;
			nt->tokLen[0]++;

			if(c == '[') {
				level++;
			}
			else if(c == ']') {
				level--;
			}
			if(level == 0) {
				return;
			}
		}
	}
}

void getQuotedStuff(nexusToken *nt)
{
	int   c, c2;
	
	// We arrive here such that the first call to fgetc() gets the opening single quote.
	c = fgetc(nt->filePtr);
	//printf("c getQuotedStuff() c=%c tokLen=%i\n", c, nt->tokLen[0]);
	if(nt->tokLen[0]) {
		printf("nextToken(), getQuotedStuff().  tokLen should be zero.\n");
		exit(1);
	}
	nt->tok[0] = c; // Save the first char.  It will always be a single quote.
	nt->tokLen[0] = 1;

	// Check that we do not have a single quote immediately following.  2 single quotes is ok, tho.
	c = fgetc(nt->filePtr);
	if(c == EOF) {
		printf("nextToken(), getQuotedStuff().  Un-matched single quote.\n");
		printf("Reached the end of the file while still in a quote.\n");
		exit(1);
	}
	else if(c == '\'') {
		c2 = fgetc(nt->filePtr);
		//printf("c getQuotedStuff() c2=%c\n", c);
		if(c2 == EOF) {
			printf("nextToken(), getQuotedStuff().\n");
			printf("Got 2 single quotes in a row, not properly within single quotes.\n");
			printf("And then the file ended.");
			exit(1);
		}
		else if(c2 == '\'') {
			// The opening single quote has been directly followed by 2 single quotes.  Thats ok.
			fseek(nt->filePtr, -2, 1);
		}
		else {
			// The opening quote was followed by a single quote and then something else.  Thats bad.
			printf("nextToken(), getQuotedStuff().\n");
			printf("Got 2 single quotes in a row, not properly within single quotes.\n");
			exit(1);
		}
	}
	else { // The opening single quote was not followed by a single quote.  Back up one space.
		fseek(nt->filePtr, -1, 1);
	}

	// We are now at position 1 in the quote.  Read in the rest.
	while(1) {
		c = fgetc(nt->filePtr);
		//printf("c getQuotedStuff() x c=%c\n", c);
		if(c == EOF) {
			printf("nextToken(), getQuotedStuff().  Un-matched single quote.\n");
			printf("Reached the end of the file while still in a quote.\n");
			exit(1);
		}
		if(c == '\n' || c == '\r') {
			printf("nextToken(), getQuotedStuff().\n");
			printf("Got a line ending while still in a quote.\n");
			printf("This can lead to trouble, as the token can exceed the buffer, so it is not allowed.\n");
			printf("Its a bug.  Sorry.\n");
			exit(1);
		}
		else if(c == '\'') {
			nt->tok[nt->tokLen[0]] = c;
			nt->tokLen[0]++;
			c2 = fgetc(nt->filePtr);
			if(c2 == EOF) {
				break;
			}
			else if(c2 == '\'') {
				nt->tok[nt->tokLen[0]] = c2;
				nt->tokLen[0]++;
				continue;
			}
			else {
				// The char following the single quote is something
				// else.  We want to get it on the next nextTok, so
				// back up one space.
				fseek(nt->filePtr, -1, 1);
				break;
			}
		}
		else {
			nt->tok[nt->tokLen[0]] = c;
			nt->tokLen[0]++;
		}
	}
	
}

void getWord(nexusToken *nt)
{
	int   c;
	int   c2;
	int   wordIsFinished = 0;

	
	while(1) {
		c = fgetc(nt->filePtr);
		//printf("    c getWord() c=%c \n", c);
		if(c == '[') {
			//printf("c getWord() about to handleComment() \n");
			handleComment(nt);
			//printf("c getWord() returned from handleComment() \n");

			if(nt->embeddedCommentLen[0]) {
				// We got a comment, and will be returning a comment.
				// First find out whether the word is finished.			
				c2 = fgetc(nt->filePtr);
				//printf("    c2 is %c\n", c2);
				wordIsFinished = 0;
				if(c2 == EOF) {
					wordIsFinished = 1;
				}
				else if(c2 == ' ' || c2 == '\n' || c2 == '\t' || c2 == '\r' ||
						c2 == '(' || c2 == ')' || c2 == '[' || c2 == ']' || c2 == '{' || 
						c2 == '}' || c2 == '\\' || c2 == '/' || c2 == ',' || c2 == ';' || 
						c2 == ':' || c2 == '=' || c2 == '*' || c2 == '\'' || c2 == '"' || 
						c2 == '`' || c2 == '+' || c2 == '-' || c2 == '<' || c2 == '>') {
					wordIsFinished = 1;
					fseek(nt->filePtr, -1, 1); // back up 1 only if we went forward 1
				}
				else {
					fseek(nt->filePtr, -1, 1); // back up 1 only if we went forward 1
				}
				if(wordIsFinished) {
					nt->savedCommentLen[0] = nt->embeddedCommentLen[0];
					nt->embeddedCommentLen[0] = 0;
				}
				else {
					nt->savedCommentLen[0] = 0;
				}

				return;
			}
			else {
				continue;
			}
		}
		if(c == EOF) {
			break;
		}
		else if(c == ' ' || c == '\n' || c == '\t' || c == '\r' ||
				c == '(' || c == ')' || c == '[' || c == ']' || c == '{' || 
				c == '}' || c == '\\' || c == '/' || c == ',' || c == ';' || 
				c == ':' || c == '=' || c == '*' || c == '\'' || c == '"' || 
				c == '`' || c == '+' || c == '-' || c == '<' || c == '>') {
			fseek(nt->filePtr, -1, 1); // back up 1 only if we went forward 1
			break;
		}
		nt->tok[nt->tokLen[0]] = c;
		nt->tokLen[0]++;
	}
}


void nexusSkipPastNextSemiColon(nexusToken *nt)
{
	int c;

	while(1) {
		c = fgetc(nt->filePtr);
		if(c == EOF) {
			printf("nexusSkipPastNextSemiColon().\n");
			printf("Reached the end of the file without finding a semicolon.\n");
			exit(1);
		}
		else if(c == '[') {
			skipComment(nt);
		}
		else if(c == ';') {
			return;
		}
	}
}


int nexusTokenCheckLineLengths(nexusToken *nt)
{
	int c;
	int  longest = 0;
	int  thisLen = 0;

	while(1) {
		c = fgetc(nt->filePtr);
		if(c == EOF) {
			return longest;
		}
		else if(c == '\n' || c == '\r') {
			if(thisLen > longest) {
				longest = thisLen;
			}
			thisLen = 0;
		}
		else {
			thisLen++;
		}
	}
}

File: p4/nexustoken.py
from __future__ import print_function
import os
import string
import sys
from p4.var import var
from p4.p4exceptions import P4Error


############################################################################
#
# Token generation stuff.
#
#       safeNextTok()  # checks for None, and dies
#       nextTok()      # may return None
#
# Handling comments:
#   - Behaviour is under the control of some variables in the var module.
#   - Visible comments are never returned.  They may be printed, depending
#     on variable settings.
#   - Command comments may be returned (depending on the variable settings).
#     While it is true that comments are not tokens, the Nexus format
#     specifies informative command comments, and so they are returned by
#     nextTok() and safeNextTok().
#
# Some differences with the old nextToken() stuff:
#   - comments within quoted words are not skipped,
#   - comments are returned in one piece
#
# Below are some possibly useful functions:
#    nexusSkipPastNextSemiColon()
#    nexusSkipPastBlockEnd()
#    nexusNextCommand()        <-- does not appear to be used anywhere
#
#
############################################################################

pieces = []  # This is used in _getWord(), where it is global.
# It needs to survive multiple calls to nextTok() and _getWord()
#wordIsFinished = False
comment = None


def safeNextTok(flob, caller=None):
    t = nextTok(flob)
    if not t:
        if caller:
            gm = ["safeNextTok(), called from %s" % caller]
        else:
            gm = ["safeNextTok()"]
        gm.append("Premature Death.")
        gm.append("Ran out of understandable things to read in nexus file.")
        raise P4Error(gm)
    else:
        return t


def nextTok(flob):
    """Returns Nexus tokens or (some) comments.

    You feed this function a flob, an open file or file-like object.

    Comments are handled under the control of a few P4 variables,
    namely

        p4.var.nexus_writeVisibleComments      # write, but do not get, all [!...]
        p4.var.nexus_getP4CommandComments      # all [&&p4 ...]
        p4.var.nexus_getWeightCommandComments  # all [&w ...]
        p4.var.nexus_getAllCommandComments     # all [&...]

    """


    global comment

    if comment:
        ret = comment
        comment = None
        return ret

    while 1:
        c = flob.read(1)
        if 0:
            if c == '\n' or c == '\r':
                print('nt %3i  c = line ending' % flob.tell())
            else:
                print('nt %3i  c = %s' % (flob.tell(), c))
        if not c:
            return None
        # if 0:
        #    if c in string.whitespace:
        #        print " c %3i  whitespace" % flob.tell()
        #    else:
        #        print " c %3i  %s" % (flob.tell(), c)
        if var.nexus_getLineEndingsAsTokens:
            if c == '\n' or c == '\r':
                return c
        if c in string.whitespace:
            continue
        elif c == '[':
            ret = _handleComment(flob, c)
            if ret:
                return ret
            else:
                continue
        elif c == "'":
            return _getQuotedStuff(flob)
        elif c in var.punctuation:
            return c
        else:
            # print("nt %3i  %s" % (flob.tell(), c))
            return _getWord(flob, c)


def _handleComment(flob, c0):
    """Do the right thing with [a nexus comment].

    The only comments we are interested in are the tree weights, eg
    [&W 1/2] and command comments, eg [& foo].  Visible comments are
    not returned, but if var.nexus_writeVisibleComments is turned on
    they are written to stdout."""

    #   nexus_writeVisibleComments      # write, but do not get, all [!...]
    #   nexus_getP4CommandComments      # all [&&p4 ...]
    #   nexus_getWeightCommandComments  # all [&w ...]
    #   nexus_getAllCommandComments     # all [&...]

    gm = ["nexustoken._handleComment()"]
    localPieces = [c0]
    # presumably c0 is '['

    c = flob.read(1)
    localPieces.append(c)
    if not c:
        # empty string, ie at the end of the flob.  level can be assumed to be
        # more than zero.
        gm.append("Reached the end while still in a comment. No comment close.")
        raise P4Error(gm)
    if c == ']':
        return None   # appears to be '[]'
    if c == '!' and var.nexus_writeVisibleComments:
        theComment = _getComment(flob, localPieces)
        print(theComment)
        return None
    elif c in ['&']:  # , '\\']:
        if var.nexus_getAllCommandComments:
            return _getComment(flob, localPieces)
        c2 = flob.read(1)
        if not c2:
            gm.append("Reached the end while still in a comment.")
            raise P4Error(gm)
        localPieces.append(c2)
        if c == '&' and c2.lower() == 'w' and var.nexus_getWeightCommandComments:
            return _getComment(flob, localPieces)
        elif c2 == '&' and var.nexus_getP4CommandComments:
            # We may need to get back here, and relative seek will not work.
            startSpot = flob.tell()
            # ask whether the next 3 characters are 'p4 ' (ie p, 4, space)
            c3 = flob.read(3)
            if len(c3) != 3 or c3.lower() != 'p4 ':
                flob.seek(startSpot)
                _skipComment(flob)
                return
            # So it is a p4 comment.  Get it.
            localPieces.append(c3)
            return _getComment(flob, localPieces)
        else:
            return _skipComment(flob)  # None, normally
    else:
        _skipComment(flob)
        return


def _skipComment(flob):
    # ie level of nested comments.  This assumes that we have already read one
    # '['.
    level = 1
    while 1:
        c = flob.read(1)
        if 0:
            if c in string.whitespace:
                print("sc %3i  whitespace" % flob.tell())
            else:
                print("sc %3i  %s" % (flob.tell(), c))
        # empty string, ie at the end of the flob.  level can be assumed to be
        # more than zero.
        if not c:
            gm = ["nexustoken._skipComment()"]
            gm.append("Reached the end while still in a comment.")
            raise P4Error(gm)
        if c == '[':
            level = level + 1
        elif c == ']':
            level = level - 1
        if level == 0:
            return


def _getComment(flob, localPieces):
    # level, ie level of nested comments.  This assumes that we have already read one '['.
    level = 1
    while 1:
        c = flob.read(1)
        localPieces.append(c)
        if 0:
            if c in string.whitespace:
                print("gc %3i  whitespace" % flob.tell())
            else:
                print("gc %3i  %s" % (flob.tell(), c))
        if not c:
            gm = ["nexustoken._getComment()"]
            gm.append("Reached the end while still in a comment.")
            raise P4Error(gm)
        if c == '[':
            level = level + 1
        elif c == ']':
            level = level - 1
        if level == 0:
            return ''.join(localPieces)


def _getWord(flob, c0):
    """Get a nexus token that isnt punctuation or a comment.

    The word might be broken by comments (the authors of the Nexus
    format were mad).  Internal command comments (eg mid[&a
    comment]dle, or aWord[&a comment at the end]) are returned before
    the enclosing word, because the comment ends before the enclosing
    word."""

    global pieces
    #global wordIsFinished
    global comment

    localPieces = [c0]
    while 1:
        prevPosn = flob.tell()
        c = flob.read(1)
        if 0:
            if not c:
                print("gw %3i  empty (position given is that of the last char)" % flob.tell())
            if c in string.whitespace:
                print("gw %3i  whitespace" % flob.tell())
            else:
                print("gw %3i  %s" % (flob.tell(), c))
        if c == '[':
            theWordPiece = ''.join(localPieces)
            # print("gw comm got theWordPiece '%s'" % theWordPiece) 
            pieces.append(theWordPiece)
            localPieces = []
            ret = _handleComment(flob, c)
            # print('gw %3i  comment=%s' % (flob.tell(), ret))

            # Here it might be useful to have wordIsFinished as a global, but here it is local.

            if ret:
                # Now we need to ask if the word is finished, or whether it
                # continues immediately after the comment.
                beforeC2Posn = flob.tell()
                # print("beforC2Posn is %i" % beforeC2Posn)
                wordIsFinished = False
                c2 = flob.read(1)
                if 0:
                    if c2 in string.whitespace:
                        print("gw %3i  c2=whitespace" % flob.tell())
                    else:
                        print("gw %3i  c2=%s" % (flob.tell(), c2))

                if c2 in string.whitespace or c2 in var.punctuation or not c2:
                    wordIsFinished = True
                if c2:
                    # back up one space only if we need to
                    if c2 in string.whitespace:
                        if var.nexus_getLineEndingsAsTokens:
                            if c2 == '\n' or c2 == '\r':
                                #print("seeking %i" % beforeC2Posn)
                                flob.seek(beforeC2Posn)
                    elif c in var.punctuation:
                        #print("seeking %i" % beforeC2Posn)
                        flob.seek(beforeC2Posn)

                if wordIsFinished:
                    # The word (ie pieces) will be returned first, then comment will be returned on the following nextTok()
                    comment = ret
                else:
                    # There is more word to come, so return the comment now.
                    return ret
            else:
                # We threw the comment away.  Look for more word.
                continue

        elif c in string.whitespace or c in var.punctuation or not c:
            if c:
                # We need to back up one (unicode) char, sometimes
                if c in string.whitespace:
                    if var.nexus_getLineEndingsAsTokens:
                        if c == '\n' or c == '\r':
                            #print("seeking %i" % prevPosn)
                            flob.seek(prevPosn)
                elif c in var.punctuation:
                    #print("seeking %i" % prevPosn)
                    flob.seek(prevPosn)
            theWordPiece = ''.join(localPieces)
            # print("gw got theWordPiece '%s'" % theWordPiece) 
            pieces.append(theWordPiece)
            break
        else:
            localPieces.append(c)
    theWord = ''.join(pieces)
    pieces = []
    return theWord


def _getQuotedStuff(flob):
    complaintHead = 'nexustoken._getQuotedStuff()'
    local_pieces = ["'"]

    # Check that we do not have a single quote immediately following.
    # 2 single quotes is ok, tho.
    c = flob.read(1)
    if c:
        local_pieces.append(c)
        if c == "'":
            c2 = flob.read(1)
            if c2:
                local_pieces.append(c2)
                if c2 == "'":
                    # The opening single quote has been directly
                    # followed by 2 single quotes. Thats ok.
                    pass
                else:
                    # The opening single quote was followed by a
                    # single quote and then something else.  Thats
                    # bad.
                    gm = [complaintHead]
                    gm.append("Got 2 single quotes in a row, not properly within single quotes.")
                    raise P4Error(gm)
            else:
                gm = [complaintHead]
                gm.append(
                    "Got 2 single quotes in a row, not properly within single quotes.")
                gm.append("And then the file ended.  Bad.")
                raise P4Error(gm)
        else:
            pass
    else:
        gm = [complaintHead]
        gm.append("File ended with an un-matched single quote.")
        raise P4Error(gm)

    # print 'At the start of the while(1) loop, the file position is %i' % flob.tell()
    while 1:
        c = flob.read(1)
        # print '! c= %s' % c
        if 0:
            if not c:
                print("gq %3i  empty (position is eof)" % flob.tell())
            elif c in string.whitespace:
                print("gq %3i  whitespace" % flob.tell())
            else:
                print("gq %3i  %s" % (flob.tell(), c))
        if c:
            local_pieces.append(c)
            if c == "'":
                posnBeforeC2 = flob.tell()
                c2 = flob.read(1)
                if c2:
                    if c2 == "'":  # ie 2 single quotes in a row
                        local_pieces.append(c2)
                        continue
                    else:
                        # It was a single quote, not followed by a single quote,
                        # so we are finished.  But in finding out that fact, we
                        # have gone too far.  So back up.
                        flob.seek(posnBeforeC2)
                        break
                    
                else:
                    # It was a closing single quote, followed by nothing.
                    break

        elif not c:
            gm = [complaintHead]
            gm.append("File ended while still in a quoted word.")
            raise P4Error(gm)
    return ''.join(local_pieces)

# (Thats it for the token generation stuff)


def nexusSkipPastNextSemiColon(flob):
    complaintHead = 'nexustoken.nexusSkipPastNextSemiColon()'
    while 1:
        c = flob.read(1)
        # print "nexusSkipPastNextSemiColon() c=%s" % c
        if c:
            if c == '[':
                _skipComment(flob)
            elif c == ';':
                return
        else:
            break
    gm = [complaintHead]
    gm.append("Every Nexus command must end in a semicolon.")
    gm.append("No semicolon was found.  Premature end of file?")
    raise P4Error(gm)


def nexusSkipPastBlockEnd(flob):
    """Read up to and including a block 'end' or 'endblock'."""
    # This should only ever be issued after a semi-colon
    gm = ['nexustoken.nexusSkipPastBlockEnd()']
    while 1:
        tok = nextTok(flob)
        # print "nexusSkipPastBlockEnd() tok=%s" % tok
        if tok:
            lowTok = tok.lower()
            if lowTok == 'end' or lowTok == 'endblock':
                tok2 = nextTok(flob)
                if not tok2 or tok2 != ';':
                    print(complaintHead)
                    gm.append("Expecting a semicolon after %s" % tok)
                    if not tok2:
                        gm.append("Got nothing.")
                    else:
                        gm.append("Got '%s'" % tok2)
                    raise P4Error(gm)
                return
            # for pathological cases where the last command is a ';' by itself.
            elif lowTok == ';':
                continue
            else:
                nexusSkipPastNextSemiColon(flob)
        else:
            break
    gm.append("Failed to find either 'end' or 'endblock'")
    gm.append("Premature end of file?")
    raise P4Error(gm)


def nexusNextCommand(flob):                           # does not appear to be used anywhere
    """Return a nexus command as a list of tokens.

    Does not return the final semi-colon."""

    toks = []
    tok = nextTok(flob)
    if not tok:  # end of file, not an error
        return None
    if tok.lower() == '#nexus':
        tok = nextTok(flob)
    while tok:
        if tok == ';':
            return toks
        toks.append(tok)
        tok = nextTok(flob)
    gm = ["nexustoken.nexusNextCommand()"]
    gm.append("Could not find a semi-colon to end the nexus command.")
    gm.append("Premature end of file?")
    raise P4Error(gm)
